### ISSUE TITLE
(Crash) Loading Classic NES Series game crash GBARunner2

### ISSUE THREAD

**Issue by Galenmacil** [Gericom/GBARunner2#118]
Hardware: DSi

GBA Rom IDs causing crash (from [DAT-o-MAtic website](https://datomatic.no-intro.org/?page=search)):

1492 - FZLE; AGB-FZLP-EUR Classic NES Series - The Legend of Zelda (USA, Europe)
1494 - FSME; AGB-FSMP-EUR Classic NES Series - Super Mario Bros. (USA, Europe)
1721 - FLBE; AGB-FLBP-EUR Classic NES Series - Zelda II - The Adventure of Link (USA, Europe)

All are proper dumps. They all work in VisualBoy Advance-M emulator.

Crash occurs when loading GBARunner2 (GBARunner2_arm7dldi_dsi.nds) via TwilightMenu++ either with DLDI ARM7 or DLDI ARM9 settings. I also tried loading GBARunner2_arm9dldi_ds.nds via a SuperCard DSTWO with the same problem.

**Comment by zenseii** [comment]
The NES Classic emulator is famous for going out of it's way to make emulation hard by being as bugged as possible. Take a read here: https://mgba.io/2014/12/28/classic-nes/

Currently the only known NES Classic ROM that runs in GBAR2 is Star Soldier.

Also, out of curiosity, why not use the NES emulator in twilightmenu++, nesDS, instead?
It will probably take a while until these ROMs become playable.
Here is the current compatibility list which is being updated to reflect the recent changes for DSi and 3DS.

**Comment by Galenmacil** [comment]
OK, Thanks for clarifying that. The main reason I would prefer using Classic NES Series on GBA instead of on a NES Emulator is because of the DSi native screen resolution (256x192). The NES, as you probably know, as a higher resolution than this on the Y plane causing either badly interpolated image or missing part of it...

It this an issue on the 2DS/3DS?

**Comment by zenseii** [comment]
The 2DS/3DS has a resolution of 400x240 so that one has no problems since the NES is 320x240. Actually it looks really good on the 3DS because of the inherent "scanlines" produced by the screen being stereoscopic 3D. They are very subtle and many might not even notice it but it does create a sort of sharpness to the pixels. This is not the case on the 2D family of systems for example but honestly it's hard to notice.

So if I got it right you would rather have the NES picture squished down to the gba resolution than the DSi one by using nesDS?

**Comment by Galenmacil** [comment]
Humm... Put it that way.

Well, the squishing on the GBA is not a purely "reduce the size of the screen vertically" method. The sprites are changed and also, possibly, some cropping is involved.

I am not asking that you add supports for these specific GBA roms. After reading the explanation on the mGBA emulator page, it became clear to me that this is not going to be simple.

Also, I just ordered myself a New 2DS XL.

Thanks for the effort you put on GBARunner2.

**Comment by zenseii** [comment]
Thanks but I'm merely a tester and have done none of the coding. Gericom deserves all the praise for that.
I hope you'll enjoy it as the N2DSXL is a very nice system in my opinion. Especially now with the latest developments the last couple of months. Considering GBAR2 now has great audio and PS1 emulation has reached very good playable levels and the first promising steps into N64 emulation have been taken.

Anyways good luck.

**Comment by cotodevel** [comment]
Hmm. I tested a patched Classic NES Series Super Mario Bros (UE) and manually forced the savetype to EEPROM in gbaemu4ds (https://bitbucket.org/Coto88/gbaemu4ds) and it actually works. (albeit bugged graphics because a GBA video mode not properly implemented)

From what I recall I emulate the EWRAM mirrors in there, so it definitely isn't a highrom issue.

It was a long time I actually played that game, lol, like 14 years ago.

**Comment by Gericom** [comment]
Yeah the use of an EWRAM mirror is one of the issues. It is not easy to solve because main memory is used for multiple purposes. There is rom data, ewram data and gbarunner2 data. I can't simply protect the entire region. That would also be very inefficient.

**Comment by cotodevel** [comment]
That memory layout is also true for gbaemu4ds. So as a workaround,I used:

https://bitbucket.org/Coto88/gbaemu4ds/src/303c49a231c573a0deb681deec89ccfe5f1b9929/hyperspeedup/arm9/source/cpumg.cpp#lines-210

and patched MPU Region Protect Settings depending on a specific game, header, or whatever.

The "trick" or "speedup" was to use multiple MPU Region Protect Settings on the uncached EWRAM, enabling the caches for I/D TCM.

**Comment by qamichele5959** [comment]
Hot mario compaign on gbar2?

**Comment by qamichele5959** [comment]
![20211013_025229](https://user-images.githubusercontent.com/88638615/137054146-f270812f-1b5f-4b52-890b-32c1bda6e341.jpg)

**Comment by zensemic59** [comment]
just use mgba3ds instead.
